go/store/datas/database_common.go

```diff
@@ -547,67 +555,192 @@ func (db *database) doUpdateWorkingSet(ctx context.Context, datasetID string, wo
 		return err
 	}
 
-	// This could loop forever, given enough simultaneous writers. BUG 2565
+	workingSetRef, err := db.WriteValue(ctx, workingSet) // will be orphaned if the tryCommitChunks() below fails
+	if err != nil {
+		return err
+	}
+
+	wsValRef, err := types.ToRefOfValue(workingSetRef, db.Format())
+	if err != nil {
+		return err
+	}
+
 	var tryCommitErr error
 	testSetFailed := false
 	for tryCommitErr = ErrOptimisticLockFailed; tryCommitErr == ErrOptimisticLockFailed && !testSetFailed; {
 		tryCommitErr = func() error {
 			db.mu.Lock()
 			defer db.mu.Unlock()
 
+			success, err := db.assertDatasetHash(ctx, datasetID, currHash)
+			if err != nil {
+				return err
+			}
+
+			if !success {
+				testSetFailed = true
+			}
+
+			currentDatasets, err := db.Datasets(ctx)
+			if err != nil {
+				return err
+			}
+
+			currentDatasets, err = currentDatasets.Edit().Set(types.String(datasetID), wsValRef).Map(ctx)
+			if err != nil {
+				return err
+			}
+
 			currentRootHash, err := db.rt.Root(ctx)
+			if err != nil {
+				return err
+			}
+
+			return db.tryCommitChunks(ctx, currentDatasets, currentRootHash)
+		}()
+	}
+
+	return tryCommitErr
+}
+
+// assertDatasetHash returns true if the hash of the dataset matches the one given. Use an empty hash for a dataset you
+// expect not to exist.
+// Typically this is called using pessimistic locking by the caller in order to implement atomic test-and-set semantics.
+func (db *database) assertDatasetHash(
+	ctx context.Context,
+	datasetID string,
+	currHash hash.Hash,
+) (bool, error) {
+
+	ds, err := db.GetDataset(ctx, datasetID)
+	if err != nil {
+		return false, err
+	}
+
+	if ds.HasHead() {
+		head, ok := ds.MaybeHead()
+		if !ok {
+			panic("no head found")
+		}
+
+		h, err := head.Hash(db.Format())
+		if err != nil {
+			return false, err
+		}
+
+		if h != currHash {
+			return false, err
+		}
+	} else {
+		if !currHash.IsEmpty() {
+			panic("No ref found for workspace " + datasetID)
```

Comment:
Is this right? Isn't this just, caller expected dataset to exist but it does not? Caller failed the optimistic lock check?

Another way to think of the question: Is there a way to remove a dataset that's going to be asserted on without risking a panic?